task/transcode.go

```diff
@@ -0,0 +1,237 @@
+package task
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/livepeer/joy4/av"
+	"github.com/livepeer/joy4/av/avutil"
+	"github.com/livepeer/joy4/format"
+	"github.com/livepeer/joy4/format/mp4"
+	"github.com/livepeer/joy4/format/ts"
+	"github.com/livepeer/livepeer-data/pkg/data"
+	"github.com/livepeer/stream-tester/model"
+	"github.com/livepeer/stream-tester/segmenter"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	segLen = 5 * time.Second
+)
+
+func init() {
+	format.RegisterAll()
+	rand.Seed(time.Now().UnixNano())
+}
+
+func TaskTranscode(tctx *TaskContext) (*data.TaskOutput, error) {
+	var (
+		ctx             = tctx.Context
+		inputPlaybackID = tctx.InputAsset.PlaybackID
+		lapi            = tctx.lapi
+	)
+
+	fullPath := videoFileName(inputPlaybackID)
+	fir, err := tctx.inputOS.ReadData(ctx, fullPath)
+	if err != nil {
+		panic(err)
+	}
+	fileInMem, err := io.ReadAll(fir.Body)
+	fir.Body.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	streamName := fmt.Sprintf("vod_%s", time.Now().Format("2006-01-02T15:04:05Z07:00"))
+	profiles := tctx.Params.Transcode.Profiles
+	stream, err := lapi.CreateStreamEx(streamName, false, nil, profiles...)
+	if err != nil {
+		return nil, err
+	}
+	defer lapi.DeleteStream(stream.ID)
+
+	fmt.Printf("Created stream id=%s name=%s\n", stream.ID, stream.Name)
+	gctx, gcancel := context.WithCancel(ctx)
+	defer gcancel()
+	segmentsIn := make(chan *model.HlsSegment)
+	if err = segmenter.StartSegmentingR(gctx, &ReaderClose{*bytes.NewReader(fileInMem)}, true, 0, 0, segLen, false, segmentsIn); err != nil {
+		return nil, err
+	}
+	var outFiles []av.Muxer
+	var outBuffers []*WriterSeeker
+	for range tctx.Params.Transcode.Profiles {
+		ws := &WriterSeeker{}
+		mp4muxer := mp4.NewMuxer(ws)
+		outFiles = append(outFiles, mp4muxer)
+		outBuffers = append(outBuffers, ws)
+	}
+	var transcoded [][]byte
+	err = nil
+out:
+	for seg := range segmentsIn {
+		if seg.Err == io.EOF {
+			break
+		}
+		if seg.Err != nil {
+			err = seg.Err
+			glog.Errorf("===> got error %v", err)
+			break
+		}
+		glog.V(model.DEBUG).Infof("Got segment seqNo=%d pts=%s dur=%s data len bytes=%d\n", seg.SeqNo, seg.Pts, seg.Duration, len(seg.Data))
+		started := time.Now()
+		transcoded, err = lapi.PushSegment(stream.ID, seg.SeqNo, seg.Duration, seg.Data)
```

Comment:
Hmm, I thought we would be able to use the `livepeer-transcode` CLI for the transcode task. Do you think there would any fundamental limitations for evolving to that? Does it support only seekable files as input? If so, would a `bytes.Reader` be enough on that case, since we are relying on the buffered file in memory anyway? Or we could also download the file to disk and then send it to the CLI, which would also have a lower footprint 🤔

Reply:
I think it was mainly 2 points:
 - Avoiding repeating the logic for tnrascoding files both on the `livepeer-transcode` and on the task-runner
 - Use the task-runner as an internal dog fooder of the `livepeer-transcode` tool. Any improvement we made for the task runner would necessarily be gained by the public users of the CLI tool as well

And if we wanted to avoid the inefficiency of spawning a separate process, we could also break up the `main` function of it just like we did for the `analyzer`, so it can be used as an internal dependency of another package by calling some `Run` function that receives all the CLI args as a parameter: 
https://github.com/livepeer/livepeer-data/blob/c87020230260e838b88223ff017bf2e56e5df0c1/cmd/analyzer/analyzer.go#L108 (in the case of the `analyzer`, would also need to move the `parseFlags` outside `Run` and receive the `cli` as an arg instead)